[File System Access article] Native File System → File System Access (#3967)
GoogleChrome/web.dev: src/site/content/en/blog/file-system-access/index.md

---
title: 'The File System Access API: simplifying access to local files'
subhead:
  The File System Access API allows web apps to read or save changes directly to files and folders
  on the user's device.
authors:
  - petelepage
  - thomassteiner
description:
  The File System Access API enables developers to build powerful web apps that interact with files
  on the user's local device, like IDEs, photo and video editors, text editors, and more. After a
  user grants a web app access, this API allows them to read or save changes directly to files and
  folders on the user's device.
date: 2019-08-20

> updated: 2020-09-28

Author: ```suggestion
updated: 2020-10-02
```

## What is the File System Access API? {: #what-is-it }

> The [File System Access API][spec] (formerly known as the Writeable Files API) enables developers to

Author: ```suggestion
The [File System Access API][spec] (formerly known as Native File System API
and prior to that Writeable Files API) enables developers to
```

## What is the File System Access API? {: #what-is-it }

The [File System Access API][spec] (formerly known as Native File System API

> and prior to that Writeable Files API) enables developers to

Author: The API was initially called Writeable Files API, then Native File System API, and finally now we settled on File System Access API.

> and prior to that Writeable Files API) enables developers to

Author: ```suggestion
and prior to that it was called Writeable Files API) enables developers to
```

`true` or `false`, you can determine if a new file or folder should be created if it doesn't exist.

```js
// In an existing directory, create a new directory named "My Documents".
const newDirectoryHandle = await existingDirectoryHandle.getDirectoryHandle('My Documents', {
  create: true,
});
// In this new directory, create a file named "My Notes.txt".
const newFileHandle = await newDirectoryHandle.getFileHandle('My Notes.txt', { create: true });
```

### Resolving the path of an item in a directory

When working with files or folders in a directory, it can be useful to resolve the path of the item
in question. This can be done with the aptly named [`resolve()`][resolve] method. For resolving, the
item can be a direct or indirect child of the directory.

```js
// Resolve the path of the previously created file called "My Notes.txt".
const path = await newDirectoryHandle.resolve(newFileHandle);
// `path` is now ["My Documents", "My Notes.txt"]
```

### Deleting files and folders in a directory

If you have obtained access to a directory, you can delete the contained files and folders with the

> [`removeEntry()`][removeentry] method, for folders optionally recursively including all subfolders
> and the therein contained files.

Author: ```suggestion
[`removeEntry()`][removeentry] method. For folders, deletion can optionally be recursive
and include all subfolders and the therein contained files.
```

```js
// Delete a file.
await directoryHandle.removeEntry('Abandoned Masterplan.txt');
// Recursively delete a folder.
await directoryHandle.removeEntry('Old Stuff', { recursive: true });
```

### Accessing the origin-private file system

The origin-private file system is a storage endpoint that, as the name suggests, is private to the
origin of the page. While browsers will typically implement this by persisting the contents of this
origin-private file system to disk somewhere, it is _not_ intended that the contents be easily user
accessible. Similarly, there is _no_ expectation that files or directories with names matching the

> names of children of the origin-private file system exist. You can operate as usual on the

Author: What this is saying is that while the browser might make it *seem* that there are files, internally—since this is an origin-private file system—the browser might store these "files" in a database or any other data structure. Essentially: what you create with this API, don't expect to find it 1:1 somewhere on the hard disk.

> names of children of the origin-private file system exist. You can operate as usual on the

Reviewer: Then please say it the way you just said it me and we'll be done.

> accessible. Similarly, there is _no_ expectation that files or directories with names matching the
> names of children of the origin-private file system exist. You can operate as usual on the

Author: ```suggestion
accessible. Similarly, there is _no_ expectation that files or directories with names matching the
names of children of the origin-private file system exist.
While the browser might make it seem that there are files, internally—since this is an origin-private file system—the browser might store these "files" in a database or any other data structure.
Essentially: what you create with this API, do *not* expect to find it 1:1 somewhere on the hard disk.
You can operate as usual on the
```

origin-private file system once you have access to the root `FileSystemDirectoryHandle`.

```js
const root = await navigator.storage.getDirectory();
// Create a new file handle.
const fileHandle = await root.getFileHandle('Untitled.txt', { create: true });
// Create a new directory handle.
const dirHandle = await root.getDirectoryHandle('New Folder', { create: true });
// Recursively remove a directory.
await root.removeEntry('Old Stuff', { recursive: true });
```

## Polyfilling

> It is not possible to completely polyfill the File System Access API methods.

Reviewer: This implies that I can polyfill its properties. Is that true?

Author: Not the properties, but the behavior. Mostly. I've outlined in the bullets what's possible and what not.

```js
// Resolve the path of the previously created file called "My Notes.txt".
const path = await newDirectoryHandle.resolve(newFileHandle);
// `path` is now ["My Documents", "My Notes.txt"]
```

### Deleting files and folders in a directory

If you have obtained access to a directory, you can delete the contained files and folders with the
[`removeEntry()`][removeentry] method. For folders, deletion can optionally be recursive
and include all subfolders and the therein contained files.

```js
// Delete a file.
await directoryHandle.removeEntry('Abandoned Masterplan.txt');
// Recursively delete a folder.
await directoryHandle.removeEntry('Old Stuff', { recursive: true });
```

### Accessing the origin-private file system

The origin-private file system is a storage endpoint that, as the name suggests, is private to the
origin of the page. While browsers will typically implement this by persisting the contents of this
origin-private file system to disk somewhere, it is _not_ intended that the contents be easily user
accessible. Similarly, there is _no_ expectation that files or directories with names matching the
names of children of the origin-private file system exist.

> While the browser might make it seem that there are files, internally—since this is an origin-private file system—the browser might store these "files" in a database or any other data structure.

Author: ```suggestion
While the browser might make it seem that there are files, internally—since this is an
origin-private file system—the browser might store these "files" in a database or any
other data structure.
```
